source/src/ble/cs_CrownstoneCentral.cpp

```diff
@@ -0,0 +1,663 @@
+	}
+	enableNotifications();
+}
+
+void CrownstoneCentral::onRead(ble_central_read_result_t& result) {
+	cs_ret_code_t retCode = result.retCode;
+	if (retCode != ERR_SUCCESS) {
+		finalizeOperation(_currentOperation, retCode);
+		return;
+	}
+
+	switch (_currentOperation) {
+		case Operation::CONNECT: {
+			ConnectSteps currentStep = static_cast<ConnectSteps>(_currentStep);
+			switch (currentStep) {
+				case ConnectSteps::SESSION_DATA: {
+					if (!finalizeStep(_currentStep, retCode)) {
+						return;
+					}
+					// Parse data that was read.
+					if (result.data.len != sizeof(session_data_t)) {
+						finalizeOperation(_currentOperation, ERR_WRONG_PAYLOAD_LENGTH);
+						return;
+					}
+					// Store the session data.
+					retCode = ConnectionEncryption::getInstance().setSessionData(*reinterpret_cast<session_data_t*>(result.data.data));
+					if (retCode != ERR_SUCCESS) {
+						finalizeOperation(_currentOperation, retCode);
+						return;
+					}
+					// Done.
+					finalizeOperation(_currentOperation, retCode);
+					break;
+				}
+				case ConnectSteps::SESSION_KEY: {
+					if (!finalizeStep(_currentStep, retCode)) {
+						return;
+					}
+					// Parse and store the session key.
+					retCode = KeysAndAccess::getInstance().setSetupKey(result.data);
+					if (retCode != ERR_SUCCESS) {
+						finalizeOperation(_currentOperation, retCode);
+						return;
+					}
+					// Read session data.
+					retCode = BleCentral::getInstance().read(_sessionDataHandle);
+					if (retCode != ERR_WAIT_FOR_SUCCESS) {
+						finalizeOperation(_currentOperation, retCode);
+						return;
+					}
+					setStep(ConnectSteps::SESSION_DATA);
+					return;
+				}
+				default:
+					finalizeOperation(_currentOperation, ERR_WRONG_OPERATION);
+					return;
+			}
+			break;
+		}
+		default: {
+			finalizeOperation(_currentOperation, ERR_WRONG_OPERATION);
+			return;
+		}
+	}
+
+}
+
+void CrownstoneCentral::onWrite(cs_ret_code_t retCode) {
+	if (retCode != ERR_SUCCESS) {
+		finalizeOperation(_currentOperation, retCode);
+		return;
+	}
+	switch (_currentOperation) {
+		case Operation::CONNECT: {
+			if (!finalizeStep(ConnectSteps::ENABLE_NOTIFICATIONS, retCode)) {
+				return;
+			}
+			readSessionData();
+			break;
+		}
+		case Operation::WRITE: {
+			if (!finalizeStep(WriteControlSteps::WRITE, retCode)) {
+				return;
+			}
+			// Wait for notifications.
+			setStep(WriteControlSteps::RECEIVE_RESULT);
+			break;
+		}
+		default: {
+			finalizeOperation(_currentOperation, ERR_WRONG_OPERATION);
+			return;
+		}
+	}
+}
+
+void CrownstoneCentral::onNotification(ble_central_notification_t& result) {
+	if (result.handle != _resultHandle) {
+		LOGw("Unexpected handle=%u", result.handle);
+		return;
+	}
+
+	// In theory, we can receive notifications at any moment.
+	// So only handle them when we expect them, and ignore them otherwise.
+	if (_currentOperation == Operation::WRITE && static_cast<WriteControlSteps>(_currentStep) == WriteControlSteps::RECEIVE_RESULT) {
```

Comment:
for readability you can negate this if and early return to win an unindent.

Reply:
Yeah, I usually do that. I was expecting another operation to be added in the future, but i guess we'll worry about that in the future.